kraken: Add new realtime trip into new mvj and vj

benoit-bst opened the pull request:
**Remains to be done** :

- [x] Add physical mode into new vj + test, **JIRA** : https://jira.kisio.org/browse/NAVP-1065
- [x] Add company into new vj + test, **DONE**

**Tests cover**
- Add new meta VJ mechanism
- Add new VJ mechanism
- Activation flags mechanism (is_realtime_add_enabled, is_realtime_add_trip_enable)
- Check impacts
- Check new journey
- Tests when you send the same GTFS-RT twice
- Test adding 2 different GTFS-RT
- Reject GTFS-RT if company id doesn't match
- Reject GTFS-RT if physical mode id doesn't match

**JIRA** 
- https://jira.kisio.org/browse/NAVP-1066
- https://jira.kisio.org/browse/NAVP-1067 (added part)

**Warning**
- now, if you can't match the _company id_ in GTFS-RT with the company inside Kraken data, there is just a log to inform you (Add inside prometheus probe is better I think). In fact, in this case, this is very dangerous to acces to the company field of the VJ _new trip_ (Example: vj->company->uri) because it is a NULL ptr... It is memory access violation behavior
**Solution** In this case, our choice is to _reject the GTFS-RT new trip_. **DONE** see test
Same technic is used for physical mode id. **DONE** see test

benoit-bst committed:
Add new trip into new mvj and vj

benoit-bst committed:
add new test with sequence

benoit-bst commented on source/kraken/apply_disruption.cpp:
```diff
@@ -208,6 +208,10 @@ struct add_impacts_visitor : public apply_impacts_visitor {
+            } else {
+                // Take the first route into the data (it is temporary).
+                // It'll replace by new method to create new default route.
+                r = pt_data.routes[0];
```
For the moment, this is the first route in the _table Routes_

benoit-bst commented on source/kraken/tests/realtime_test.cpp:
```diff
@@ -2740,3 +2740,299 @@ BOOST_AUTO_TEST_CASE(should_get_correct_base_stop_time_with_lollipop_II) {
+
+
+
+    // Recall the same new trip.
```
You have to see _test logs_ to glimpse the behavior
```
...
[19-02-26 15:15:38,317] [TRACE] [] - realtime trip update received realtime.cpp:537  
[19-02-26 15:15:38,317] [DEBUG] [] - Vehicle journey founded : vj_new_trip realtime.cpp:566  
[19-02-26 15:15:38,317] [DEBUG] [] - Creating disruption realtime.cpp:366  
[19-02-26 15:15:38,317] [DEBUG] [] - Deleting disruption: feed-1 apply_disruption.cpp:664  
[19-02-26 15:15:38,317] [DEBUG] [] - Deleting impact: feed-1 apply_disruption.cpp:652  
[19-02-26 15:15:38,317] [DEBUG] [] - we are going to cleanup the vj vj_new_trip:modified:0:feed-1 type.cpp:501  
[19-02-26 15:15:38,317] [DEBUG] [] - feed-1 deleted apply_disruption.cpp:655  
[19-02-26 15:15:38,317] [DEBUG] [] - disruption feed-1 deleted apply_disruption.cpp:675  
[19-02-26 15:15:38,317] [TRACE] [] - a trip has been changed realtime.cpp:421  
[19-02-26 15:15:38,317] [TRACE] [] - Adding stop time into impact realtime.cpp:424  
[19-02-26 15:15:38,317] [DEBUG] [] - Trip effect : additional service realtime.cpp:506  
[19-02-26 15:15:38,317] [DEBUG] [] - Disruption added realtime.cpp:525  
[19-02-26 15:15:38,317] [DEBUG] [] - applying disruption: feed-1 apply_disruption.cpp:681  
[19-02-26 15:15:38,317] [TRACE] [] - Adding impact: feed-1 apply_disruption.cpp:517  
[19-02-26 15:15:38,317] [TRACE] [] - Start to add impact feed-1 on object vj_new_trip apply_disruption.cpp:116  
[19-02-26 15:15:38,317] [TRACE] [] - modifying vj_new_trip apply_disruption.cpp:206 
...
```
the function that do this job
```
void delete_disruption(const std::string& disruption_id, ...) apply_disruption.cpp:660
```

benoit-bst replied:
As @kadhikari said, we want to see into the test a difference, to show if the second GTFS-RT replace the first.
To do that, I change in the second GTFS-RT only the stops time schedule.

benoit-bst committed:
Add tests for company

benoit-bst committed:
highlight company field issue

kinnou02 commented on source/kraken/realtime.cpp:
```diff
@@ -57,8 +57,9 @@ static bool is_handleable(const transit_realtime::TripUpdate& trip_update){
     // enum in gtfs-rt proto to express this idea
-    if (trip_update.trip().schedule_relationship() == transit_realtime::TripDescriptor_ScheduleRelationship_SCHEDULED
-            && trip_update.stop_time_update_size()) {
+    if ((trip_update.trip().schedule_relationship() == transit_realtime::TripDescriptor_ScheduleRelationship_SCHEDULED ||
```
I think that checking the features flags (`is_realtime_add_enabled` and `is_realtime_add_trip_enabled`) should be done here.

benoit-bst replied:
I think you have right, Wa can add the flags logic inside this function (is_handleable).
Flags logic
https://github.com/CanalTP/navitia/pull/2689/files/bc83c537ac0a8fd0e60e119d38651f48bc72fbd8#diff-a1c1532aa9f44bfd3b8a77e208515e23R544

kinnou02 commented on source/tests/utils_test.h:
```diff
@@ -21,7 +21,7 @@ inline void handle_realtime_test(const std::string& id,
                                  const type::Data& data,
                                  std::unique_ptr<navitia::routing::RAPTOR>& raptor) {
-    navitia::handle_realtime(id, timestamp, trip_update, data, true, true);
+    navitia::handle_realtime(id, timestamp, trip_update, data,true);
```
spacing
```suggestion
    navitia::handle_realtime(id, timestamp, trip_update, data, true);
```

benoit-bst committed:
kinnou's remarks

benoit-bst committed:
test: second GTFS-RT is now different

kadhikari commented on source/kraken/tests/realtime_test.cpp:
```diff
@@ -2740,3 +2740,317 @@ BOOST_AUTO_TEST_CASE(should_get_correct_base_stop_time_with_lollipop_II) {
+    BOOST_CHECK_EQUAL(res.response_type(), pbnavitia::NO_SOLUTION);
+    BOOST_CHECK_EQUAL(res.journeys_size(), 0);
+
+
```
You can delete some empty lines

kadhikari commented on source/kraken/tests/realtime_test.cpp:
```diff
@@ -2740,3 +2740,317 @@ BOOST_AUTO_TEST_CASE(should_get_correct_base_stop_time_with_lollipop_II) {
+    BOOST_REQUIRE_EQUAL(pt_data.meta_vjs.size(), 1);
+    BOOST_REQUIRE_EQUAL(pt_data.meta_vjs.exists("vj_new_trip"), false);
+
+
```
Same here

kadhikari commented on source/kraken/tests/realtime_test.cpp:
```diff
@@ -2740,3 +2740,317 @@ BOOST_AUTO_TEST_CASE(should_get_correct_base_stop_time_with_lollipop_II) {
+    BOOST_CHECK_EQUAL(vj->base_validity_pattern()->days, year("00000000"));
+    BOOST_CHECK_EQUAL(vj->adapted_validity_pattern()->days, year("00000000"));
+    BOOST_CHECK_EQUAL(vj->rt_validity_pattern()->days, year("00000001"));
+
```
If i understand, for a new added trip we have:
- base vj with pattern 0s
- adapted vj with pattern 0s and 
- rt vj with pattern 0s and 1 once.

benoit-bst replied:
when we have an Add Trip, we create a VJ within one new meta VJ.
Indeed, this VJ has no **base** and **adapted** validity pattern but only a **RT** validity pattern.
In this example, we have only one disrupted day, that's the reason for _0000000**1**_

kadhikari commented on source/kraken/tests/realtime_test.cpp:
```diff
@@ -2740,3 +2740,317 @@ BOOST_AUTO_TEST_CASE(should_get_correct_base_stop_time_with_lollipop_II) {
+    // Recall the same new trip.
+    // The old "new trip" is replace by the new, but the meta VJ is the same (don't create).
+    // The old impact is suppress too
+    navitia::handle_realtime("feed-1", timestamp, new_trip, *b.data, true, true);
```
Is it possible to modify some informations in the trip to be added for the second time and verify those modifications?

benoit-bst committed:
we reject GTFS-RT if company_id doesn't match + tests

benoit-bst committed:
Empty space

pbougue commented on source/kraken/realtime.cpp:
```diff
@@ -281,6 +247,69 @@ static bool is_added_stop_time(const transit_realtime::TripUpdate& trip_update)
+    // Case 2: the trip is not cancelled, but modified (ex: the train's arrival is delayed or the train's
+    // departure is brought forward), we check the size of stop_time_update because we can't find a proper
+    // enum in gtfs-rt proto to express this idea
+    if ((trip_update.trip().schedule_relationship() == transit_realtime::TripDescriptor_ScheduleRelationship_SCHEDULED ||
```
As you use this `if` multiple times (3 in this PR, maybe more?), is it possible to extract that, for example in a `bool TripUpdate.is_circulating()` ? It's the most expressive I can think of.

benoit-bst replied:
It don't mind to have this cond. But if it scratches too much I can do something like that
```
is_circulating(trip_update.trip().schedule_relationship()) {

     if trip_update.trip().schedule_relationship() == trt::TripDescriptor_ScheduleRelationship_SCHEDULED || trip_update.trip().schedule_relationship() == trt::TripDescriptor_ScheduleRelationship_ADDED)&& trip_update.stop_time_update_size())
        return true
    else 
        return false
}
```

pbougue replied:
Yep, you can do it as an external function as well (would take trip_update as argument, as you use also stoptimeupdate).
My main point is to mutualize maintenance of this statement and promote future reuse.
Having a "one-liner" or not is just codestyle (I prefer "one-liner" like in the current `if` statement).

benoit-bst replied:
it's okay for me

pbougue commented on source/kraken/apply_disruption.cpp:
```diff
@@ -499,7 +503,8 @@ static bool is_modifying_effect(nt::disruption::Effect e) {
                   nt::disruption::Effect::DETOUR,
-                  nt::disruption::Effect::REDUCED_SERVICE});
+                  nt::disruption::Effect::REDUCED_SERVICE,
+                  nt::disruption::Effect::ADDITIONAL_SERVICE});
```
:+1:

pbougue commented on source/kraken/realtime.cpp:
```diff
@@ -281,6 +247,69 @@ static bool is_added_stop_time(const transit_realtime::TripUpdate& trip_update)
 }
 
+static bool is_handleable(const transit_realtime::TripUpdate& trip_update,
+                          nt::PT_Data& pt_data,
```
This should probably be `const` (then maybe a const access has to be added, but it'd be safer)

benoit-bst replied:
:+1:

pbougue commented on source/kraken/realtime.cpp:
```diff
@@ -281,6 +247,69 @@ static bool is_added_stop_time(const transit_realtime::TripUpdate& trip_update)
+            pt_data.companies_map.find(trip_update.trip().GetExtension(kirin::company_id)) == pt_data.companies_map.end()) {
+            LOG4CPLUS_DEBUG(log, "Trip company "
+                    << trip_update.trip().GetExtension(kirin::company_id)
+                    <<  " don't exists: ignoring trip id "
```
:gb: : "do`es`n't exist`:` ignoring trip `update` id "

benoit-bst replied:
thanks

pbougue commented on source/kraken/realtime.cpp:
```diff
@@ -281,6 +247,69 @@ static bool is_added_stop_time(const transit_realtime::TripUpdate& trip_update)
+        // check if company id exists
+        if (trip_update.trip().HasExtension(kirin::company_id) &&
+            pt_data.companies_map.find(trip_update.trip().GetExtension(kirin::company_id)) == pt_data.companies_map.end()) {
+            LOG4CPLUS_DEBUG(log, "Trip company "
```
minor: "Trip company `id` "

pbougue commented on source/kraken/realtime.cpp:
```diff
@@ -492,10 +522,16 @@ create_disruption(const std::string& id,
 
-            trip_effect = get_calculated_trip_effect(most_important_stoptime_status);
+            if (trip_update.trip().schedule_relationship() == trt::TripDescriptor_ScheduleRelationship_ADDED) {
+                // Yeah, that's quite hardcoded...
```
:smile: we may add a mapping method to cover those.
Maybe use/adapt the `get_calculated_trip_effect` from below?

benoit-bst replied:
We can't use this method but we can rename it like `get_calculated_stop_trip_effect`

benoit-bst replied:
I prefer to keep this method only to find the effect inside stop time if it is needed

kadhikari replied:
As `get_calculated_trip_effect` will be removed soon, we may leave as it is

pbougue commented on source/kraken/realtime.cpp:
```diff
@@ -528,31 +564,21 @@ void handle_realtime(const std::string& id,
-        return;
+    // Get or create meta VJ
+    navitia::type::MetaVehicleJourney* meta_vj = nullptr;
+    if (! data.pt_data->meta_vjs.exists(trip_update.trip().trip_id())) {
```
We add it only if it's an ADDED trip, I think.
Otherwise it should be rejected.

benoit-bst replied:
Good idea. I think it is preferable too

pbougue commented on source/kraken/realtime.cpp:
```diff
@@ -528,31 +564,21 @@ void handle_realtime(const std::string& id,
+    if (! data.pt_data->meta_vjs.exists(trip_update.trip().trip_id())) {
+        LOG4CPLUS_DEBUG(log, "unknown vehicle journey, create Meta VJ " << trip_update.trip().trip_id());
+        meta_vj = data.pt_data->meta_vjs.emplace(trip_update.trip().trip_id());
+        meta_vj->tz_handler = data.pt_data->tz_manager.get_first_timezone();
```
Let's add a TODO and list it in what should be properly managed (JIRA?).

benoit-bst replied:
To create a Meta VJ you need to add a new in _pt_data->meta_vjs_ (with _emplace_ method) and create a new _meta_vj->tz_handler_.
What do you want to doc and where ?

pbougue replied:
If I understand correctly, the TZ you chose is maybe not the correct one, we have to pick the right one (don't know how we have to choose it: main one from the coverage?).
So I'd add a `TODO: pick a meaningful TZ` in code, and keep track of it in a JIRA ticket with everything we should choose more carefully than "additional service" for example :wink:

benoit-bst replied:
okay I'll add a TODO comment for that

pbougue commented on source/kraken/realtime.cpp:
```diff
@@ -528,31 +564,21 @@ void handle_realtime(const std::string& id,
-        return;
+    // Get or create meta VJ
+    navitia::type::MetaVehicleJourney* meta_vj = nullptr;
+    if (! data.pt_data->meta_vjs.exists(trip_update.trip().trip_id())) {
```
We could save a search and still use get_mut (test if nullptr).

benoit-bst replied:
for me, it just a reading logic. It is more clear to have exists and 2 method for read (get) and create (create). 
When you don't know the code this is more complex to understand with only _get_mut_... that was my case.

On the other hand, if you want to take into account your remark on `We add it only if it's an ADDED trip, I think.
Otherwise it should be rejected.`

You need to have something like that
```
bool exists = data.pt_data->meta_vjs.exists(trip_update.trip().trip_id()
if (!exist && new trip = added){
   Create meta VJ with emplace
}
else if (!exist && new trip != added) {
   // Don't want to create a meta VJ
   return
}
else {
    // meta vj exist
   get_mut() to retrieve it
}
```

pbougue reviewed (commented):
Did not review tests.

pbougue commented on source/kraken/apply_disruption.cpp:
```diff
@@ -208,6 +208,10 @@ struct add_impacts_visitor : public apply_impacts_visitor {
                                                          meta.production_date);
             if (! r && ! mvj->get_base_vj().empty()) {
                 r = mvj->get_base_vj().at(0)->route;
+            } else {
```
This should be a `else if (! r)`.
You may reorganize the code, but you get the spirit :wink:

benoit-bst replied:
I let you change that when you merged the work. This is your pluging place :hear_no_evil:

benoit-bst committed:
add physical mode management in New RT

benoit-bst commented:
Since I have updated with physical mode id (Manage physical mode into new VJ), the **data version number** has been incremented. Beware of that

benoit-bst committed:
test add physical mode check inside new vj

benoit-bst committed:
create is_circulating() + reject meta VJ none without Added type

benoit-bst committed:
test: fix typo and comments

pbougue commented on source/kraken/apply_disruption.cpp:
```diff
@@ -240,8 +244,11 @@ struct add_impacts_visitor : public apply_impacts_visitor {
+                if (!impact->physical_mode_id.empty()) {
+                    nu::make_map_find(pt_data.physical_modes_map, impact->physical_mode_id)
+                        .if_found([&vj](navitia::type::PhysicalMode* p){ vj->physical_mode = p; })
+                        .if_not_found([&](){ LOG4CPLUS_WARN(log, "[disruption] Associate physical mode into new VJ. Physical mode doesn't exist with id : " << impact->physical_mode_id); });
```
In that case it will crash later on raptor build.
I think we should protect a bit (even if done earlier in handleable() ):
* `impact->physical_mode_id` if found
* else `pt_data.physical_modes[0];`

benoit-bst replied:
It should be a poor protection but useful to avoid runtime crash

kinnou02 replied:
I agree with @pbougue

benoit-bst replied:
On my way to add this protection, but don't forget that, if the key is not finded inside the map, we reject directly the GTFS-RT - > https://github.com/CanalTP/navitia/pull/2689/files#diff-a1c1532aa9f44bfd3b8a77e208515e23R301

I'll add the same protection for company :tulip:

pbougue commented on source/kraken/realtime.cpp:
```diff
@@ -281,6 +256,76 @@ static bool is_added_stop_time(const transit_realtime::TripUpdate& trip_update)
+                    << trip_update.vehicle().GetExtension(kirin::physical_mode_id)
+                    <<  " doesn't exist: ignoring trip update id "
+                    << trip_update.trip().trip_id());
+            return false;
```
Maybe we can fallback on `physical_mode:Coach` if it exists (most common case considering SNCF or other transport providers).
Maybe we can do this fallback in apply_disruption directly.

benoit-bst replied:
For the moment if we don't match with the key, as we said, we reject. What is the problem with that ?

pbougue replied:
:heavy_check_mark: Discussed and agreed on another feed (company+physical_mode management)

pbougue commented on source/kraken/realtime.cpp:
```diff
@@ -281,6 +256,76 @@ static bool is_added_stop_time(const transit_realtime::TripUpdate& trip_update)
+                    << trip_update.trip().GetExtension(kirin::company_id)
+                    <<  " doesn't exist: ignoring trip update id "
+                    << trip_update.trip().trip_id());
+            return false;
```
The situation I can imagine here (same for physical_mode below) is:
We run this kraken version + old kirin version (rollback for a problem or because IRE is back...).
Then Kraken would just reject all TripUpdates as they do not implem company_id or physical_mode_id, no?

So I think we should reject those only when we absolutely need it (trip add?). But it should be mostly backward compatible.

pbougue replied:
This is probably the case for GTFS-RT already (company and physical_mode = `""`).
We have to plug Sherbrooke GTFS-RT on this PR to check the behavior **#QA**

pbougue replied:
Tested plugging Sherbrooke's GTFS-RT on kirin: it looks like it's working (checked on departures).
To me, it still looks dangerous to over-reject here.

benoit-bst replied:
To be clear, we check the _company id_ or the _physical mode id_ only if the GTFS-RT contains the extension. 
I think Sherbrooke's GTFS-RT doesn't send the extension ?

pbougue replied:
Issue is the same for company and physical_mode.
True, I checked that for Sherbrooke Kirin doesn't send the extension, so it works.
Still, on feeds that are not adding trips, I'm not at all confident with the idea of being stricter than we used to be.
We agreed on a solution orally:
* rejecting strictly only on trip-adds

benoit-bst commented:
Jenkins, test this please !

kadhikari commented on source/kraken/realtime.cpp:
```diff
@@ -528,31 +580,27 @@ void handle_realtime(const std::string& id,
+                << trip_update.trip().trip_id());
         return;
+    } else {
+        LOG4CPLUS_DEBUG(log, "Vehicle journey founded : " << trip_update.trip().trip_id());
```
typo : founded -> found

kinnou02 commented on source/type/data.cpp:
```diff
@@ -61,7 +61,7 @@ namespace pt = boost::posix_time;
 namespace navitia { namespace type {
 
-const unsigned int Data::data_version = 69; //< *INCREMENT* every time serialized data are modified
+const unsigned int Data::data_version = 70; //< *INCREMENT* every time serialized data are modified
```
In fact you do not need to change `data_version` as the `data.nav.lz4` doesn't contain any `Impact`.
The serializer of `Impact` is only used during a clone, inside a process, but never to be persisted.

pbougue replied:
OK, I'm fine with both.
But it's also saving a lot of time at deploy saving re-bina on every coverage to stay on the same version... So I'll vote to go back to 69 on that.

benoit-bst replied:
Nice catch @kinnou02. Go back to `Data::data_version = 69`

kinnou02 reviewed (changes requested):
Change proposed by @pbougue must be implemented before merging this PR.

Also, It will be nice to add a jormungandr integration test that add a vj, and to do a few requests that use it. just to catch the most obvious bug.
Integration tests should be run with the `add_*` flags set to true

pbougue commented on source/tests/utils_test.h:
```diff
@@ -62,15 +62,24 @@ inline transit_realtime::TripUpdate
 make_delay_message(const std::string& vj_uri,
```
To be renamed `make_ntfs_realtime_message` (assuming that NTFS-RT is a real thing :smile: ).
These are not just delay anymore and these are different from Chaos realtime messages.

benoit-bst replied:
Same thing, this is not my creation, Maybe in an other PR with title "improving realtime test readability" ?
It is just because the PR to already heavy
> JIRA ticket coming up for "test refactor"

benoit-bst replied:
JIRA : https://jira.kisio.org/browse/NAVP-1228

pbougue commented on source/kraken/tests/realtime_test.cpp:
```diff
@@ -2740,3 +2740,380 @@ BOOST_AUTO_TEST_CASE(should_get_correct_base_stop_time_with_lollipop_II) {
+                 ("stop_point:D", "09:30"_t);
+
+    // Add company
+    std::string comp_name = "comp_name";
```
You can probably use the build_helper on all those: https://github.com/CanalTP/navitia/blob/dev/source/ed/build_helper.cpp#L201-L238
Or if it's simpler: https://github.com/CanalTP/navitia/blob/dev/source/ed/build_helper.cpp#L666

benoit-bst replied:
The fact is, there is no method to create only simple Company or Physical_mode like Stop_time...
I have just created my own way because I want to control all process.
But I think, this is a good remark to create a JIRA ticket within the technical debt to improve the build_helper

There is a lot of methods to add :tulip:

pbougue replied:
To be managed through JIRA ticket test refactor (+ build_helper improvement)

benoit-bst replied:
JIRA : https://jira.kisio.org/browse/NAVP-1228

pbougue commented on source/kraken/tests/realtime_test.cpp:
```diff
@@ -2740,3 +2740,380 @@ BOOST_AUTO_TEST_CASE(should_get_correct_base_stop_time_with_lollipop_II) {
+    navitia::routing::RAPTOR raptor(*(b.data));
+    ng::StreetNetwork sn_worker(*b.data->geo_ref);
+
+    auto compute = [&](const char* datetime, const std::string& from, const std::string& to) {
```
`datetime` can be a timestamp, no? Don't know about the type exactly, as readability is the goal here I agree.
At least no `char*` ;)

benoit-bst replied:
add a std::string if you prefer. This case is used in all realtime_test.cpp file.
We can do even better, like extract the _compute(...)_ function so that use in all code. But not in this PR, it is already heavy changes.

pbougue replied:
:+1: changed to string
Mutualizing the method in test_utils to come in a JIRA ticket on test refactor

benoit-bst replied:
JIRA : https://jira.kisio.org/browse/NAVP-1228

pbougue commented on source/kraken/tests/realtime_test.cpp:
```diff
@@ -2740,3 +2740,380 @@ BOOST_AUTO_TEST_CASE(should_get_correct_base_stop_time_with_lollipop_II) {
+
+    // Recall the same new trip.
+    // The old "new trip" is replace by the new, but the meta VJ is the same (don't create).
+    // The old impact is suppress too
```
`deleted`

pbougue commented on source/kraken/tests/realtime_test.cpp:
```diff
@@ -2740,3 +2740,380 @@ BOOST_AUTO_TEST_CASE(should_get_correct_base_stop_time_with_lollipop_II) {
+    BOOST_REQUIRE_EQUAL(pt_data.meta_vjs.exists("vj_new_trip_2"), false);
+
+
+    // If the physical mode id doesn't exist inside the data, we reject teh new trip
```
typo: `the`

pbougue commented on source/kraken/tests/realtime_test.cpp:
```diff
@@ -2740,3 +2740,380 @@ BOOST_AUTO_TEST_CASE(should_get_correct_base_stop_time_with_lollipop_II) {
     BOOST_CHECK_EQUAL(stoptime_rltm_D.get_base_stop_time(), &vj->stop_time_list[4]);
 }
+
+BOOST_AUTO_TEST_CASE(add_new_trip) {
```
If it's possible it would be better to split, using `BOOST_FIXTURE_TEST_CASE`.

Minor: maybe even in a new file.

benoit-bst replied:
For tests strongness, I need to follow cases, this is the reason why it is a long long test. thanks to that, we can monitor all counter example
```
BOOST_REQUIRE_EQUAL(pt_data.vehicle_journeys.size(), 1);
// Do tests
BOOST_REQUIRE_EQUAL(pt_data.vehicle_journeys.size(), 2);
etc
```

pbougue replied:
Suggested some places to separate if you think it's worth it.
I resolve this feed but use that boost fixture if you split tests ;)

pbougue reviewed (commented):
@kinnou02 as for the `add_*` flags set to true, I think it's what you mentionned: https://github.com/CanalTP/navitia/pull/2683/files#diff-ca32dae0e1de944fd72a73775b54cfe1R753

- [ ] Although there are no integration tests adding trip, I agree we need it : something probably inspired from https://github.com/CanalTP/navitia/pull/2581/files#diff-ca32dae0e1de944fd72a73775b54cfe1R744

Most of the TODO from this previous PR should be covered before merging it I think: https://github.com/CanalTP/navitia/pull/2679#pullrequestreview-205697712

> :heavy_check_mark: JIRA Ticket for integration tests **DONE**

pbougue commented on source/kraken/tests/realtime_test.cpp:
```diff
@@ -2740,3 +2740,380 @@ BOOST_AUTO_TEST_CASE(should_get_correct_base_stop_time_with_lollipop_II) {
+    BOOST_REQUIRE_EQUAL(pt_data.meta_vjs.exists("vj_new_trip"), false);
+
+
+    // New call with all activation parameter = true
```
We may split tests here (separate tests on flags above from the other ones)

pbougue commented on source/kraken/tests/realtime_test.cpp:
```diff
@@ -2740,3 +2740,380 @@ BOOST_AUTO_TEST_CASE(should_get_correct_base_stop_time_with_lollipop_II) {
+    BOOST_CHECK_EQUAL(res.impacts(0).impacted_objects(0).impacted_stops(3).departure_status(), pbnavitia::StopTimeUpdateStatus::ADDED);
+
+
+    // If the company id doesn't exist inside the data, we reject the new trip
```
Those tests on rejection based on company and physical_mode could be independent.
(I know that depending on how it's rejected, those tests may change a bit)

benoit-bst committed:
split tests + lenient with trip with id + downgrade version

benoit-bst committed:
update utils

benoit-bst commented:
To sum up :

- Update the Utils lib (push force was the way... :hear_no_evil: sorry)
- The _data version number_ is always the same (thanks @kinnou02 )
- Refactoring realtime_test will be in an other PR for better readability (JIRA : https://jira.kisio.org/browse/NAVP-1228)
- Integration tests will be inside in an other PR for better readability (JIRA : https://jira.kisio.org/browse/NAVP-1194)
- Add dataset into VJ. 
Currently, when we create new VJ we don't create or link with a dataset... _Dataset = nullptr_. it seems to work so far but it is not very safe. I added this one now (Test are added too https://github.com/CanalTP/navitia/pull/2689/files#diff-5675d2990ed70bf3b28b8586e5e566eaR2915)
- Added new protection to reject GTFS-RT. 
If _company key_ or _physical mode key_ doesn't match with data, while we have the Extension, we reject only if it is a GTFS-RT ADDED Type. More secure to work with old Kirin version (-> https://github.com/CanalTP/navitia/pull/2689/files#diff-a1c1532aa9f44bfd3b8a77e208515e23R292). 
It works to have a company id with SCHEDULED (https://github.com/CanalTP/navitia/pull/2689/files#diff-5675d2990ed70bf3b28b8586e5e566eaR241)
- Split tests because @pbougue doen't support :rose: 

notes :
in an other work, we'll need new methods to create dataset, contributor, company and physical mode by **default**, for a new VJ. Currently, we use pt_data.companies[0], pt_data.datasets[0], ...
example in the code (->https://github.com/CanalTP/navitia/pull/2689/files#diff-c234f9d9583465696c5f3311313759d0R293 ) : 
it is more preferable to have something like that
```
vj->company = create_company("default_name");  
vj->dataset = create_dataset("default_name");  
...
```
all work in the commit (https://github.com/CanalTP/navitia/pull/2689/commits/d5affca2fa9c269c36aa032ed34ab7643c545150)

pbougue commented on source/kraken/apply_disruption.cpp:
```diff
@@ -220,18 +220,59 @@ struct add_impacts_visitor : public apply_impacts_visitor {
+
+            // Add company
             if (!impact->company_id.empty()) {
+                LOG4CPLUS_TRACE(log, "kikou " << impact->company_id);
```
Not sure you wanna keep this trace :smile: 
If you want to keep it, Alex's pseudo is "ki`nn`ou" :wink: 
(don't know about the others also, but they may help)

benoit-bst replied:
hahaha my famous **kikou message** :speak_no_evil:

pbougue commented on source/kraken/apply_disruption.cpp:
```diff
@@ -220,18 +220,59 @@ struct add_impacts_visitor : public apply_impacts_visitor {
+                        // for protection, use the companies[0]
+                        // TODO : Create default company
+                        vj->company = pt_data.companies[0];
+                        LOG4CPLUS_WARN(log, "[disruption] Associate company into new VJ. Company doesn't exist with id : " << impact->company_id); });
```
If you keep this trace, maybe add : "Associate `random` company `to` new VJ"

pbougue commented on source/kraken/apply_disruption.cpp:
```diff
@@ -220,18 +220,59 @@ struct add_impacts_visitor : public apply_impacts_visitor {
+                if (! mvj->get_base_vj().empty()) {
+                    vj->company = mvj->get_base_vj().at(0)->company;
+                } else {
+                    // for protection, use the companies[0]
```
I think you can name and use the same functor used in `if_not_found`

pbougue commented on source/kraken/apply_disruption.cpp:
```diff
@@ -220,18 +220,59 @@ struct add_impacts_visitor : public apply_impacts_visitor {
+                        // for protection, use the physical_modes[0]
+                        // TODO : Create default physical mode
+                        vj->physical_mode = pt_data.physical_modes[0];
+                        LOG4CPLUS_WARN(log, "[disruption] Associate physical mode into new VJ. Physical mode doesn't exist with id : " << impact->physical_mode_id); });
```
Same remark on trace.

pbougue commented on source/kraken/apply_disruption.cpp:
```diff
@@ -220,18 +220,59 @@ struct add_impacts_visitor : public apply_impacts_visitor {
+                if (! mvj->get_base_vj().empty()) {
+                    vj->physical_mode = mvj->get_base_vj().at(0)->physical_mode;
+                } else {
+                    // for protection, use the physical_modes[0]
```
Same remark on functor.

pbougue commented on source/kraken/apply_disruption.cpp:
```diff
@@ -220,18 +220,59 @@ struct add_impacts_visitor : public apply_impacts_visitor {
+                        LOG4CPLUS_WARN(log, "[disruption] Associate company into new VJ. Company doesn't exist with id : " << impact->company_id); });
+            } else {
+                if (! mvj->get_base_vj().empty()) {
+                    vj->company = mvj->get_base_vj().at(0)->company;
```
:+1: Nice!

pbougue commented on source/kraken/apply_disruption.cpp:
```diff
@@ -240,16 +281,17 @@ struct add_impacts_visitor : public apply_impacts_visitor {
+                // for protection, use the datasets[0]
+                // TODO : Create default data set
+                vj->dataset = pt_data.datasets[0];
+                LOG4CPLUS_WARN(log, "[disruption] Associate dataset into new VJ doesn't work, take first data set by default");
```
minor: Maybe just "take `random dataset`" (one word for dataset)

pbougue commented on source/kraken/tests/realtime_test.cpp:
```diff
@@ -2741,60 +2742,94 @@ BOOST_AUTO_TEST_CASE(should_get_correct_base_stop_time_with_lollipop_II) {
+        b.data->pt_data->contributors.push_back(contributor);
+        b.data->pt_data->contributors_map[contributor->uri] = contributor;
+
+        // Add data set
```
minor: `dataset` in one word

pbougue reviewed (commented):
A few code style remarks left.

I'm OK for the rest.
Agreed also on the work left, using the same kind of get_or_create for everything in pt_data :yum:

benoit-bst commented:
Jenkins, test this please !

kinnou02 commented on source/kraken/realtime.cpp:
```diff
@@ -46,38 +46,15 @@ namespace navitia {
+{
+    if ((trip_update.trip().schedule_relationship() == transit_realtime::TripDescriptor_ScheduleRelationship_SCHEDULED ||
+         trip_update.trip().schedule_relationship() == transit_realtime::TripDescriptor_ScheduleRelationship_ADDED)
+         && trip_update.stop_time_update_size()) {
```
```suggestion
         && trip_update.stop_time_update_size() > 1) {
```
If there is only one `stop_time_update` we can discard it

benoit-bst replied:
rather that ?
```
 && trip_update.stop_time_update_size() >= 1) {
```
if the GTFS-RT contains one stop_time we can process ?

pbougue replied:
@kinnou02 you mean that we are supposed to get trip_updates that are complete trips, and a trip with only one stop is useless, so we should get rid of it, I guess.
So he means `> 1`.

Do we do that when we read NTFS?
This would remove the disruption also (not sure it's bad though).
We could as well reject them in Kirin.
Both are fine to me.

kinnou02 replied:
A vj with only one stop isn't a vj, it's garbage :)

benoit-bst replied:
I am all right with that. I take your suggest

benoit-bst committed:
comments more consistent

benoit-bst commented:
I have updated Log messages as mentionned comments.

@pbougue I don't understand your remark `I think you can name and use the same functor used in if_not_found`
Please comment with suggested change if you have a clear idea or resolve conversation. Thank you in advance

I rebased the dev, so a _push forced_ was needed... the commit to read for new changes (https://github.com/CanalTP/navitia/pull/2689/commits/9c57ba2b49d865cdb13673fa9401f2a2f47f5248)

kinnou02 committed:
Update source/kraken/realtime.cpp

A vj with only one stop isn't a vj

Co-Authored-By: benoit-bst <32099204+benoit-bst@users.noreply.github.com>

benoit-bst commented:
Since we have merged the @kinnou02 's change, `&& trip_update.stop_time_update_size() > 1) {`
integration tests failed. 
Several tests create a mock with just one _UpdateStopTime_ like this
```  
 # New disruption with a deleted stop_time recently added at stop_point:stopC                                                                                                                                                                                           
self.send_mock(                                                                                                                                                                                                                                                        
               "vjA",                                                                                                                                                                                                                                                             
               "20120614",                                                                                                                                                                                                                                                        
               'modified',                                                                                                                                                                                                                                                        
               [                                                                                                                                                                                                                                                                  
                   UpdatedStopTime(                                                                                                                                                                                                                                               
                       "stop_point:stopC",                                                                                                                                                                                                                                        
                       arrival_delay=0,                                                                                                                                                                                                                                           
                       departure_delay=0,                                                                                                                                                                                                                                         
                       arrival=tstamp("20120614T080104"),                                                                                                                                                                                                                         
                       departure=tstamp("20120614T080104"),                                                                                                                                                                                                                       
                       message='stop_time deleted',                                                                                                                                                                                                                               
                       arrival_skipped=True,
                   )
               ],
              disruption_id='delete_stop_time',
)
```
Is it realistic to receive that sort of GTFS-RT, with only one _Stop Time_ ? @pbougue @kadhikari

benoit-bst commented:
After discussion with @pbougue, it seems to be more comfortable to accept a **GTFS-RT** with only **one StopTimeUpdate** even if it s not VJ compliant. 
There are many reasons
- We can Update just one _Stop time_, the format doesn't prohibit this practice, even if it is always a UpdateStopTime repetition for all _GTFS-RT_ that we receive
- Because many tests are lenient with that sort of GTFS-RT
- Maybe if one day we receive a _partial/differential tripUpdate_

I hope you agree with that. :bird:

benoit-bst committed:
revert Update source/kraken/realtime.cpp

nonifier reviewed (approved):
:bullettrain_side::train::train::monkey::train::train:

pbougue reviewed (dismissed):
Looks like you downgraded source/utils : please place it on CanalTP's `master` branch.
There was a mix-up :wink: 
Currently navitia's `dev` branch is on util's `master` branch, and this PR shouldn't change that.

> feel free to dismiss this review once this is managed

benoit-bst committed:
update utils

benoit-bst committed:
Fix building data after a realtime add

We don't wnat to sort index.
So a simple build_raptor() is realistic after handle_realtime calling.

benoit-bst committed:
take into account of effect

benoit-bst commented:
I updated the PR to work with **TripUpdate.effect**
I needed to update some Utests with effect.

**QA Part** :
- Kirin : _last version_
- Jormun : _last version_
- Kraken : _this PR version_

1. launch COTS - add new trip
File : trip_add_paris_marseille.json
Result : **GOOD**

2. launch COTS - Sequence 09 case 1
Delay All stop points with a medium delay (10min) + add 2 new stop point at the end.
File : trip_seq9_01_delays_and_new_stop_points.json
Result : **GOOD**

3. launch COTS - Sequence 09 case 2
Change delays (60 mins) of the last 5 stop point.
File : trip_seq9_02_update_delays.json
Result : **GOOD**

4. launch COTS - Sequence 09 case 3
Remove the 2 added stop point.
File : trip_seq9_03_delays_and_delete_stop_points.json
Result : **GOOD**

I think it is enough. If you want to test more, please do (rebase on this PR for the kraken version)

pbougue commented on source/kraken/realtime.cpp:
```diff
@@ -46,14 +46,44 @@ namespace navitia {
 
 namespace nd = type::disruption;
 
+static bool is_cancelled(const transit_realtime::TripUpdate& trip_update)
```
for consistency: is_canceled`_trip` would be nicer (and to avoid any misunderstanding with canceled stop)

pbougue commented on source/kraken/realtime.cpp:
```diff
@@ -46,14 +46,44 @@ namespace navitia {
+    }
+}
+
 static bool is_circulating(const transit_realtime::TripUpdate& trip_update)
```
minor: maybe is_circulating`_trip` also here

pbougue commented on source/kraken/realtime.cpp:
```diff
@@ -540,12 +580,7 @@ create_disruption(const std::string& id,
-            } else {
-                trip_effect = get_calculated_trip_effect(most_important_stoptime_status);
-            }
+            trip_effect = get_calculated_trip_effect(most_important_stoptime_status);
```
I think you still have to manage `trip_effect = nt::disruption::Effect::ADDITIONAL_SERVICE;` (it's not managed from stops), no?

benoit-bst replied:
ADDTIONAL_SERVICE is a "main effect". It is manage here
https://github.com/CanalTP/navitia/pull/2689/files#diff-a1c1532aa9f44bfd3b8a77e208515e23R589

pbougue commented on source/kraken/tests/realtime_test.cpp:
```diff
@@ -2350,7 +2353,8 @@ BOOST_AUTO_TEST_CASE(add_modify_and_delete_new_stop_time_in_the_trip) {
                     RTStopTime("stop_point:C", "20171101T0900"_pts),
-            });
+            },
+            transit_realtime::Alert_Effect::Alert_Effect_DETOUR);
```
It's more `Alert_Effect_REDUCED_SERVICE` here, no?
(And it's probably more MODIFIED_SERVICE above on line 2314)

benoit-bst replied:
It is a `Effect DETOUR`
https://github.com/CanalTP/navitia/pull/2689/files#diff-5675d2990ed70bf3b28b8586e5e566eaR2377

pbougue replied:
It's a leftover from an old way of choosing effect (still in `get_calculated_trip_effect` which is deprecated), it has changed in Kirin since then.
It only affects this test, so it's fine to let it now. For consistency, I will just change it in my PR that focuses on cleaning tests.

pbougue commented on source/kraken/tests/realtime_test.cpp:
```diff
@@ -2390,7 +2394,8 @@ BOOST_AUTO_TEST_CASE(add_modify_and_delete_new_stop_time_in_the_trip) {
                     RTStopTime("stop_point:C", "20171101T0900"_pts),
-            });
+            },
+            transit_realtime::Alert_Effect::Alert_Effect_DETOUR);
```
Same, it's not a detour I think it's more `Alert_Effect_REDUCED_SERVICE`

benoit-bst replied:
It is a `Effect DETOUR`
https://github.com/CanalTP/navitia/pull/2689/files#diff-5675d2990ed70bf3b28b8586e5e566eaR2413

pbougue replied:
Same as above.

pbougue commented on source/tests/utils_test.h:
```diff
@@ -62,11 +62,12 @@ inline transit_realtime::TripUpdate
 make_delay_message(const std::string& vj_uri,
         const std::string& start_date,
         const std::vector<RTStopTime>& delayed_time_stops,
+        const transit_realtime::Alert_Effect effect = transit_realtime::Alert_Effect::Alert_Effect_MODIFIED_SERVICE,
```
In Kirin, MODIFIED_SERVICE is more for stop adds, so I'd default on SIGNIFICANT_DELAYS (or UNKNOWN_EFFECT).
https://github.com/CanalTP/kirin/blob/51e728e10b2009ccd756b29cbd1c59cd66d33e13/kirin/core/types.py#L83-L96

benoit-bst replied:
You should probably right. I think `SIGNIFICANT_DELAYS` is the best to have a default value for test.

pbougue commented on source/tests/utils_test.h:
```diff
@@ -79,7 +80,7 @@ make_delay_message(const std::string& vj_uri,
     // https://github.com/CanalTP/chaos-proto/blob/master/gtfs-realtime.proto#L459
     trip->set_start_date(start_date);
-    trip->set_schedule_relationship(effect);
+    trip->set_schedule_relationship(transit_realtime::TripDescriptor_ScheduleRelationship_SCHEDULED);
```
I think we should remove it if possible (or compute it from effect but...)

benoit-bst replied:
Yes you've right, I can remove it now

pbougue commented on source/kraken/realtime.cpp:
```diff
@@ -496,6 +584,7 @@ create_disruption(const std::string& id,
         } else {
```
Just a reminder here because it's an "outdated" comment.
Above:
I think you still have to manage `trip_effect = nt::disruption::Effect::ADDITIONAL_SERVICE;` (it's not managed from get_calculated_trip_effect(stops) ), no?
We should see it in test, maybe.

benoit-bst replied:
ADDTIONAL_SERVICE is a "main effect". It is manage here
https://github.com/CanalTP/navitia/pull/2689/files#diff-a1c1532aa9f44bfd3b8a77e208515e23R589

benoit-bst committed:
Change Effect by default into test set creator

Commit:
Manage full back-to-normal feeds (UNKNOWN_EFFECT) + tests update

benoit-bst committed:
fix disruption inside journeys

benoit-bst commented:
The disruption disappeared from Journeys. I fixed it and added a test.
**QA**
Now send `trip_delay_9580_tgv.json` then `trip_base_9580_tgv.json` works :fireworks:

benoit-bst commented:
**To sum up the QA part**

New trip
- Add new trip - _DONE_
- Add Delays on new trip - _DONE_
- Add new Stop Time on new trip - _DONE_
- Remove one Stop Time on new trip - _DONE_
- Back to normal after all - _DONE_

Sequence of cases with stop time
- Test case 09 - flux 1 - _DONE_
- Test case 09 - flux 2 - _DONE_
- Test case 09 - flux 3 - _DONE_

Simple case with stop time
- trip_delay_9580_tgv.json - _DONE_
- trip_base_9580_tgv.json - _DONE_

benoit-bst committed:
finally Unknown effect disruption is only in disruption API

benoit-bst commented:
It will be the last commit

We have passed a lof of time to speak about disruption inside Journeys API when there is a **Back to Normal** GTFS-RT in input. 
It turns out, the last behaviour was, only print the disruption with `UNKNOWN_EFFECT` inside the _disruption API_ (not in _Journeys_).
The associate PR was https://github.com/CanalTP/navitia/pull/2199
 
Now, Utest and integration tests highlight that.

benoit-bst commented:
Artemis is **GREEN** :tada:
